iree/turbine/kernel/wave/codegen/emitter.py

```diff
@@ -193,55 +197,95 @@ def add_emitter_subs(
     return dynamics
 
 
-_emulate_ceildiv = True
+_emulate_ceildiv = bool(int(environ.get("WAVE_EMULATE_CEILDIV", 1)))
+_use_affine_expr = bool(int(environ.get("WAVE_USE_AFFINE_EXPR", 0)))
 
 _Rational = namedtuple("_Rational", ["numerator", "denominator"])
+_ApplyExpr = namedtuple("_ApplyExpr", ["expr", "args"])
```

Comment:
Could you instead make this a class and in the class verify that all the args are of type index? If they are of type index, you can return an ApplyExpr, otherwise you can return the value. This way the parts that can leverage affine apply will and others won't and you won't need to throw an exception.

Reply:
fixed in different way